array_api_strict/_data_type_functions.py

```diff
@@ -204,35 +204,31 @@ def result_type(
     # required by the spec rather than using np.result_type. NumPy implements
     # too many extra type promotions like int64 + uint64 -> float64, and does
     # value-based casting on scalar arrays.
-    A = []
+    dtypes = []
     scalars = []
     for a in arrays_and_dtypes:
-        if isinstance(a, Array):
-            a = a.dtype
+        if isinstance(a, DType):
+            dtypes.append(a)
+        elif isinstance(a, Array):
+            dtypes.append(a.dtype)
         elif isinstance(a, (bool, int, float, complex)):
             scalars.append(a)
-        elif isinstance(a, np.ndarray) or a not in _all_dtypes:
-            raise TypeError("result_type() inputs must be array_api arrays or dtypes")
-        A.append(a)
-
-    # remove python scalars
-    B = [a for a in A if not isinstance(a, (bool, int, float, complex))]
+        else:
+            raise TypeError(
+                "result_type() inputs must be Array API arrays, dtypes, or scalars"
+            )
 
-    if len(B) == 0:
+    if not dtypes:
```

Comment:
This got me off at a first glance, but yeah it's OK because `dtypes` contains both arrays and dtypes.
Is fine, even if resembles a bit a snippet I recently saw in somebody's going through the "intro to python" online course: `mytuple = []`.

Reply:
The `dtypes` list contains exclusively dtypes though.